### DISCUSSION TITLE
Loss calculation for NER while training

### DISCUSSION THREAD

**Discussion by sudarshan-koirala** [explosion/spaCy#9605] [Help: Model Advice] [answered]
Hi, I am training a NER model with transformer. From this discussion [9129](https://github.com/explosion/spaCy/discussions/9128) it is clear that it is using `cpu_log_loss` for the primary loss calculation.  Also from this discussion [4094](https://github.com/explosion/spaCy/issues/4094) and [3634](https://github.com/explosion/spaCy/issues/3634), it is mentioned that for a predicted entity to be counted as correct it has to be a perfect match with a tuple from the gold annotation. There are no partial matches or weighting.

My question now is , to calculate loss while training, does the predicted entity has to be a perfect match with a tuple from the gold annotation to be counted as correct ?

**Comment by honnibal** [comment, marked as answer]
The topic is a bit complicated because the parser relies on a transformation of the tasks into a state sequence. What the parser actually models (and predicts) is, given the current state in a state machine, what set of actions will result in no new errors being introduced?

For the purposes of the calculation of what is an "error", the NER model does assume that only exact match counts, with no partial credit for overlap. In technical terms this is really the _oracle_ of the transition system, which can be described separately from the loss function. This translates as follows.

Let's say we have the sentence `The Ozarks is a region.`, and the gold standard has the entity `(0, 2, LOC)`.  We initialize our state machine with the state `(0, [])`: it's at word zero and there's no entities recognized. The right transition here is one which will open a new `LOC` entity and move forward to the next word. But imagine we get this wrong, and predict an action that just transitions us forward, without opening the entity. So what's optimal at `Ozarks`? The NER's oracle assumes that the `(0, 2, LOC)` entity is lost and now cannot be recovered. An action which opens (or, opens and closes) a new entity `(1, 2, LOC)` will add an incorrect entity, which will make the analysis worse. So the optimal action is to just move forward to the next word.

Other oracles for the NER could be defined and used, but the implementation needs to be very fast, so this is hard-coded in Cython. It's also a pretty technical topic, because the oracle needs to be well-behaved in certain respects. It needs to decompose properly, so that there's always a distinct transition to "blame" for a given error in the output.
